handbook/engineering/testing.md

```diff
@@ -2,11 +2,87 @@
 
 This file documents how we test code at Sourcegraph.
 
+## Philosophy
 
-## Percy
+We rely on automated testing to ensure the quality of our product.
+
+Any addition or change to our codebase should be covered by an appropriate amount of automated tests to ensure that:
+
+1. Our product works as intended when we ship it to customers.
+1. Our product continues to work and doesn't accidentally break as we make changes to our product over time.
+
+Engineers should budget for writing tests when making iteration plans.
+
+## Flaky tests
+
+We do not tolerate flaky tests of any kind. Any engineer that sees a flaky test should immediately:
+
+1. Open a PR to disable the flaky test.
+1. Open an issue to re-enable the flaky test, assign it to the most likely owner, and add it to the current release milestone.
+
+If the build or test infrastructure itself is flaky, then [open an issue](https://github.com/sourcegraph/sourcegraph/issues/new?labels=team/distribution) and notify the [distribution team](distribution/index.md#contact).
+
+## Unit tests
+
+Unit tests test individual functions in our codebase and are the most desirable kind of test to write.
+
+Benefits:
+
+- They are usually very fast to execute because slow operations can be mocked.
+- They are the easiest tests to write, debug, and maintain because the code under test is small.
+- They only need to run on changes that touch code which could make the test fail, which makes CI faster and minimizes the impact of any flakiness.
+
+Tradeoffs:
+
+- They don't verify our systems are wired up correctly end-to-end.
+
+## Integration tests
+
+Integration tests test the behavior of a subset of our entire system to ensure that subset of our system is wired up correctly.
+
+Benefits:
+
+- To the extent that fewer systems are under test compared to e2e tests, they are faster to run, easier to debug, have clearer ownership, and less vulnerable to flakiness.
+- They only need to run on changes that touch code which could make the test fail, which makes CI faster and minimizes the impact of any flakiness.
+
+Tradeoffs:
+
+- They don't verify our systems are wired up correctly end-to-end.
+- They are not as easy to write as unit tests.
+
+Examples:
+
+- Tests that call our search API to test the behavior of our entire search system.
+- Tests that validate the behavior of complex UI flows in the browser while mocking out all network requests so no backend is required.
```

Comment:
_complex_ UI flows will probably fall into the category that are _hard_ to test with integration tests, because a complex UI flow usually means there is state change on the server involved, which is difficult to replicate on the client without reimplementing the logic of the server inside tests (a mini in-memory backend), which then becomes time-consuming to maintain with every backend update. I'd actually say "complex" UI flows mark the verge to where e2e tests are more suitable. We hopefully will find some solution for this in the webapp, but this may not be the strongest example for where integration tests shine.

Possible alternative:

```suggestion
- Tests that validate UI behavior in the browser while mocking out all network requests so no backend is required.
```